File: src/detect/strategy.ts
import type { DetectionResult, DetectionSignal, StrategyType } from "./types";
import { CONFIG } from "./types";

const SCORE = {
  REQUIRE_LOAD: 8,
  LOAD_PATTERN: 2,
  REQUIRE_SETUP: 6,
  SETUP_OPTIONS: 2,
  COLORSCHEME_USAGE: 4,
  VIM_G_GLOBALS: 3,
  FILE_STRATEGY: 2,
  COLORS_VIM_ONLY: 6,
  LUA_MODULE_WITH_COLORS: 4,
  COLORS_LUA_ONLY: 5,
  PLUGIN_WITH_LUA: 3,
  LUA_NO_COLORS: 4,
  BONUS_LOAD_OVER_SETUP: 2,
  BONUS_SETUP_OVER_COLORSCHEME: 3,
} as const;

const RE = {
  REQUIRE_LOAD: /require\(["'][^"']+["']\)\.load\s*\(/i,
  LOAD_PATTERN: /\.load\s*\(\s*{?/i,
  REQUIRE_CALL: /require\(/i,
  REQUIRE_SETUP: /require\(["'][^"']+["']\)\.setup\s*\(/i,
  SETUP_OPTIONS: /setup\s*\(\s*{[\s\S]*?}\s*\)/i,
  COLORSCHEME_CMD: /:?colorscheme\s+[a-z0-9_.-]+/i,
  VIM_CMD_COLORSCHEME: /vim\.cmd\s*\(\s*["']colorscheme\s+[a-z0-9_.-]+["']\s*\)/i,
  VIM_CMD_DOT_COLORSCHEME: /vim\.cmd\.colorscheme\s*\(\s*["'][a-z0-9_.-]+["']\s*\)/i,
  VIM_G_GLOBAL: /let\s+g:[a-z_]+\s*=/i,
  BACKGROUND_MODE: /background\s*=\s*["'](dark|light)["']/i,
  CUSTOM_ORDERING: /before\s+loading|after\s+loading|must\s+set\s+global/i,
  LUA_MODULE: /^lua\/[^/]+\/init\.lua$/i,
  LUA_SINGLE: /^lua\/[^/]+\.lua$/i,
  COLORS_LUA: /^colors\/.+\.lua$/i,
  COLORS_VIM: /^colors\/.+\.vim$/i,
  PLUGIN_LUA: /^plugin\/.+\.lua$/i,
} as const;

const DEFAULT_TALLY: Record<StrategyType, number> = {
  setup: 0,
  load: 0,
  colorscheme: 0,
  file: 0,
  unknown: 0,
};

function createTally(): Record<StrategyType, number> {
  return { ...DEFAULT_TALLY };
}

function computeDetection(signals: DetectionSignal[]): { detected: StrategyType; confidence: number } {
  const tally = createTally();
  for (const s of signals) tally[s.strategy] += s.score;

  if (tally.load > 0 && tally.setup > 0 && tally.load >= tally.setup) {
    tally.load += SCORE.BONUS_LOAD_OVER_SETUP;
  }
  if (tally.setup > 0 && tally.colorscheme > 0) {
    tally.setup += SCORE.BONUS_SETUP_OVER_COLORSCHEME;
  }

  const ranked = (Object.entries(tally) as Array<[StrategyType, number]>)
    .filter(([k]) => k !== "unknown")
    .sort((a, b) => b[1] - a[1]);

  const top = ranked[0] ?? ["unknown", 0];
  const second = ranked[1] ?? ["unknown", 0];

  if (top[1] === 0) {
    return { detected: "unknown", confidence: 0 };
  }

  const delta = Math.max(0, top[1] - second[1]);
  const confidence = Math.min(1, top[1] / 10 + delta / 10);

  return { detected: top[0], confidence };
}

export function detectFromText(readme: string): DetectionResult {
  const signals: DetectionSignal[] = [];
  const text = readme;
  const lower = readme.toLowerCase();

  if (RE.REQUIRE_LOAD.test(text)) {
    signals.push({ strategy: "load", score: SCORE.REQUIRE_LOAD, reason: "README contains require(...).load(...)" });
  }
  if (RE.LOAD_PATTERN.test(text) && RE.REQUIRE_CALL.test(text)) {
    signals.push({ strategy: "load", score: SCORE.LOAD_PATTERN, reason: "README shows .load() pattern" });
  }

  if (RE.REQUIRE_SETUP.test(text)) {
    signals.push({ strategy: "setup", score: SCORE.REQUIRE_SETUP, reason: "README contains require(...).setup(...)" });
  }
  if (RE.SETUP_OPTIONS.test(text)) {
    signals.push({ strategy: "setup", score: SCORE.SETUP_OPTIONS, reason: "README shows setup({...}) options block" });
  }

  if (RE.COLORSCHEME_CMD.test(text)) {
    signals.push({ strategy: "colorscheme", score: SCORE.COLORSCHEME_USAGE, reason: "README shows :colorscheme usage" });
  }
  if (RE.VIM_CMD_COLORSCHEME.test(text)) {
    signals.push({ strategy: "colorscheme", score: SCORE.COLORSCHEME_USAGE, reason: 'README shows vim.cmd("colorscheme ...")' });
  }
  if (RE.VIM_CMD_DOT_COLORSCHEME.test(text)) {
    signals.push({ strategy: "colorscheme", score: SCORE.COLORSCHEME_USAGE, reason: "README shows vim.cmd.colorscheme(...)" });
  }

  if (RE.VIM_G_GLOBAL.test(text) && !RE.REQUIRE_CALL.test(text)) {
    signals.push({ strategy: "colorscheme", score: SCORE.VIM_G_GLOBALS, reason: "README shows vim.g globals without require()" });
  }

  if (RE.BACKGROUND_MODE.test(text) && RE.COLORSCHEME_CMD.test(text)) {
    signals.push({ strategy: "file", score: SCORE.FILE_STRATEGY, reason: "README suggests mode-dependent setup + colorscheme" });
  }
  if (RE.CUSTOM_ORDERING.test(lower)) {
    signals.push({ strategy: "file", score: SCORE.FILE_STRATEGY, reason: "README suggests custom init ordering" });
  }

  const { detected, confidence } = computeDetection(signals);

  const needsSourceInspection = detected === "unknown" || confidence < CONFIG.HIGH_CONFIDENCE_THRESHOLD;

  return { detected, confidence, signals, needsSourceInspection };
}

export type FileTreeItem = { path: string; type: string };

function hasMatchingPath(paths: string[], regex: RegExp): boolean {
  return paths.some((p) => regex.test(p));
}

export function inspectSource(files: FileTreeItem[]): Partial<DetectionResult> {
  const filePaths = files.filter((t) => t.type === "blob").map((t) => t.path);

  const hasLuaModule = hasMatchingPath(filePaths, RE.LUA_MODULE) || hasMatchingPath(filePaths, RE.LUA_SINGLE);
  const hasColorsLua = hasMatchingPath(filePaths, RE.COLORS_LUA);
  const hasColorsVim = hasMatchingPath(filePaths, RE.COLORS_VIM);
  const hasPluginDir = hasMatchingPath(filePaths, RE.PLUGIN_LUA);

  const signals: DetectionSignal[] = [];

  if (hasColorsVim && !hasLuaModule && !hasColorsLua) {
    signals.push({ strategy: "colorscheme", score: SCORE.COLORS_VIM_ONLY, reason: "Repo has colors/*.vim without Lua module" });
  }

  if (hasLuaModule && hasColorsLua) {
    signals.push({ strategy: "setup", score: SCORE.LUA_MODULE_WITH_COLORS, reason: "Repo has Lua module + colors/*.lua" });
  }

  if (hasColorsLua && !hasLuaModule) {
    signals.push({ strategy: "colorscheme", score: SCORE.COLORS_LUA_ONLY, reason: "Repo has colors/*.lua without Lua module" });
  }

  if (hasPluginDir && hasLuaModule) {
    signals.push({ strategy: "setup", score: SCORE.PLUGIN_WITH_LUA, reason: "Repo has plugin/ dir + Lua module" });
  }

  if (hasLuaModule && !hasColorsLua && !hasColorsVim) {
    signals.push({ strategy: "load", score: SCORE.LUA_NO_COLORS, reason: "Repo has Lua module without colors/" });
  }

  if (signals.length === 0) {
    signals.push({ strategy: "unknown", score: 0, reason: "No clear signals from source" });
  }

  const { detected, confidence } = computeDetection(signals);

  return { detected, confidence, signals };
}

File: internal/theme/strategy.go
package theme

import (
	"math"
	"regexp"

	"github.com/raulcorreia7/theme-browser-registry/internal/gh"
)

// Ported verbatim from src/detect/strategy.ts. StrategyType constants live in
// internal/detect; here strategies are plain strings.

// Signal is one detection signal: a strategy vote with a score and reason.
type Signal struct {
	Strategy string  `json:"strategy"`
	Score    float64 `json:"score"`
	Reason   string  `json:"reason"`
}

// DetectionResult is the full outcome of DetectFromText.
type DetectionResult struct {
	Detected              string   `json:"detected"`
	Confidence            float64  `json:"confidence"`
	Signals               []Signal `json:"signals"`
	NeedsSourceInspection bool     `json:"needsSourceInspection"`
}

// PartialDetection is the InspectSource outcome (no inspection flag).
type PartialDetection struct {
	Detected   string   `json:"detected"`
	Confidence float64  `json:"confidence"`
	Signals    []Signal `json:"signals"`
}

const (
	highConfidenceThreshold = 0.9
	minConfidenceThreshold  = 0.5
)

var score = map[string]float64{
	"requireLoad":               8,
	"loadPattern":               2,
	"requireSetup":              6,
	"setupOptions":              2,
	"colorschemeUsage":          4,
	"vimGGlobals":               3,
	"fileStrategy":              2,
	"colorsVimOnly":             6,
	"luaModuleWithColors":       4,
	"colorsLuaOnly":             5,
	"pluginWithLua":             3,
	"luaNoColors":               4,
	"bonusLoadOverSetup":        2,
	"bonusSetupOverColorscheme": 3,
}

var re = map[string]*regexp.Regexp{
	"requireLoad":       regexp.MustCompile(`(?i)require\(["'][^"']+["']\)\.load\s*\(`),
	"loadPattern":       regexp.MustCompile(`(?i)\.load\s*\(\s*{?`),
	"requireCall":       regexp.MustCompile(`(?i)require\(`),
	"requireSetup":      regexp.MustCompile(`(?i)require\(["'][^"']+["']\)\.setup\s*\(`),
	"setupOptions":      regexp.MustCompile(`(?i)setup\s*\(\s*\{[\s\S]*?\}\s*\)`),
	"colorschemeCmd":    regexp.MustCompile(`(?i):?colorscheme\s+[a-z0-9_.-]+`),
	"vimCmdColorscheme": regexp.MustCompile(`(?i)vim\.cmd\s*\(\s*["']colorscheme\s+[a-z0-9_.-]+["']\s*\)`),
	"vimCmdDotScheme":   regexp.MustCompile(`(?i)vim\.cmd\.colorscheme\s*\(\s*["'][a-z0-9_.-]+["']\s*\)`),
	"vimGGlobal":        regexp.MustCompile(`(?i)let\s+g:[a-z_]+\s*=`),
	"backgroundMode":    regexp.MustCompile(`(?i)background\s*=\s*["'](dark|light)["']`),
	"customOrdering":    regexp.MustCompile(`(?i)before\s+loading|after\s+loading|must\s+set\s+global`),
	"luaModule":         regexp.MustCompile(`(?i)^lua/[^/]+/init\.lua$`),
	"luaSingle":         regexp.MustCompile(`(?i)^lua/[^/]+\.lua$`),
	"colorsLua":         regexp.MustCompile(`(?i)^colors/.+\.lua$`),
	"colorsVim":         regexp.MustCompile(`(?i)^colors/.+\.vim$`),
	"pluginLua":         regexp.MustCompile(`(?i)^plugin/.+\.lua$`),
}

// computeDetection sums per-strategy scores, applies the two bonuses, and
// derives confidence as min(1, top/10 + delta/10). Ties keep the TS tally
// insertion order (setup, load, colorscheme, file).
func computeDetection(signals []Signal) (string, float64) {
	tallyOrder := []string{"setup", "load", "colorscheme", "file"}
	tally := map[string]float64{}
	for _, k := range tallyOrder {
		tally[k] = 0
	}
	for _, s := range signals {
		tally[s.Strategy] += s.Score
	}

	if tally["load"] > 0 && tally["setup"] > 0 && tally["load"] >= tally["setup"] {
		tally["load"] += score["bonusLoadOverSetup"]
	}
	if tally["setup"] > 0 && tally["colorscheme"] > 0 {
		tally["setup"] += score["bonusSetupOverColorscheme"]
	}

	type rankedEntry struct {
		key string
		val float64
	}
	ranked := make([]rankedEntry, 0, len(tallyOrder))
	for _, k := range tallyOrder { // stable order preserved by construction
		if tally[k] != 0 {
			ranked = append(ranked, rankedEntry{k, tally[k]})
		}
	}
	// stable insertion sort by score desc
	for i := 1; i < len(ranked); i++ {
		for j := i; j > 0 && ranked[j].val > ranked[j-1].val; j-- {
			ranked[j], ranked[j-1] = ranked[j-1], ranked[j]
		}
	}

	var top, second rankedEntry
	if len(ranked) > 0 {
		top = ranked[0]
	}
	if len(ranked) > 1 {
		second = ranked[1]
	}

	if top.val == 0 {
		return "unknown", 0
	}

	delta := math.Max(0, top.val-second.val)
	confidence := math.Min(1, top.val/10+delta/10)
	return top.key, confidence
}

// DetectFromText ports src/detect/strategy.ts detectFromText.
func DetectFromText(readme string) DetectionResult {
	var signals []Signal

	hasRequireCall := re["requireCall"].MatchString(readme)
	colorschemeCmd := re["colorschemeCmd"].MatchString(readme)

	if re["requireLoad"].MatchString(readme) {
		signals = append(signals, Signal{"load", score["requireLoad"], "README contains require(...).load(...)"})
	}
	if re["loadPattern"].MatchString(readme) && hasRequireCall {
		signals = append(signals, Signal{"load", score["loadPattern"], "README shows .load() pattern"})
	}

	if re["requireSetup"].MatchString(readme) {
		signals = append(signals, Signal{"setup", score["requireSetup"], "README contains require(...).setup(...)"})
	}
	if re["setupOptions"].MatchString(readme) {
		signals = append(signals, Signal{"setup", score["setupOptions"], "README shows setup({...}) options block"})
	}

	if colorschemeCmd {
		signals = append(signals, Signal{"colorscheme", score["colorschemeUsage"], "README shows :colorscheme usage"})
	}
	if re["vimCmdColorscheme"].MatchString(readme) {
		signals = append(signals, Signal{"colorscheme", score["colorschemeUsage"], `README shows vim.cmd("colorscheme ...")`})
	}
	if re["vimCmdDotScheme"].MatchString(readme) {
		signals = append(signals, Signal{"colorscheme", score["colorschemeUsage"], "README shows vim.cmd.colorscheme(...)"})
	}

	if re["vimGGlobal"].MatchString(readme) && !hasRequireCall {
		signals = append(signals, Signal{"colorscheme", score["vimGGlobals"], "README shows vim.g globals without require()"})
	}

	if re["backgroundMode"].MatchString(readme) && colorschemeCmd {
		signals = append(signals, Signal{"file", score["fileStrategy"], "README suggests mode-dependent setup + colorscheme"})
	}
	if re["customOrdering"].MatchString(readme) {
		signals = append(signals, Signal{"file", score["fileStrategy"], "README suggests custom init ordering"})
	}

	detected, confidence := computeDetection(signals)
	if signals == nil {
		signals = []Signal{}
	}
	return DetectionResult{
		Detected:              detected,
		Confidence:            confidence,
		Signals:               signals,
		NeedsSourceInspection: detected == "unknown" || confidence < highConfidenceThreshold,
	}
}

func hasMatchingPath(paths []string, pattern *regexp.Regexp) bool {
	for _, p := range paths {
		if pattern.MatchString(p) {
			return true
		}
	}
	return false
}

// InspectSource ports src/detect/strategy.ts inspectSource.
func InspectSource(files []gh.TreeItem) PartialDetection {
	var paths []string
	for _, t := range files {
		if t.Type == "blob" {
			paths = append(paths, t.Path)
		}
	}

	hasLuaModule := hasMatchingPath(paths, re["luaModule"]) || hasMatchingPath(paths, re["luaSingle"])
	hasColorsLua := hasMatchingPath(paths, re["colorsLua"])
	hasColorsVim := hasMatchingPath(paths, re["colorsVim"])
	hasPluginDir := hasMatchingPath(paths, re["pluginLua"])

	var signals []Signal

	if hasColorsVim && !hasLuaModule && !hasColorsLua {
		signals = append(signals, Signal{"colorscheme", score["colorsVimOnly"], "Repo has colors/*.vim without Lua module"})
	}
	if hasLuaModule && hasColorsLua {
		signals = append(signals, Signal{"setup", score["luaModuleWithColors"], "Repo has Lua module + colors/*.lua"})
	}
	if hasColorsLua && !hasLuaModule {
		signals = append(signals, Signal{"colorscheme", score["colorsLuaOnly"], "Repo has colors/*.lua without Lua module"})
	}
	if hasPluginDir && hasLuaModule {
		signals = append(signals, Signal{"setup", score["pluginWithLua"], "Repo has plugin/ dir + Lua module"})
	}
	if hasLuaModule && !hasColorsLua && !hasColorsVim {
		signals = append(signals, Signal{"load", score["luaNoColors"], "Repo has Lua module without colors/"})
	}

	if len(signals) == 0 {
		signals = append(signals, Signal{Strategy: "unknown", Score: 0, Reason: "No clear signals from source"})
	}

	detected, confidence := computeDetection(signals)
	return PartialDetection{Detected: detected, Confidence: confidence, Signals: signals}
}

// MinConfidenceThreshold exposes the detect CONFIG threshold for the stage.
const MinConfidenceThreshold = minConfidenceThreshold
